src/background/uri-info.js

```diff
@@ -1,25 +1,72 @@
 import settings from './settings';
 
+const BLOCKLIST = [
+  { urlProperty: 'protocol', value: 'chrome:' },
+  { urlProperty: 'hostname', value: 'facebook.com' },
+  { urlProperty: 'hostname', value: 'www.facebook.com' },
+  { urlProperty: 'hostname', value: 'mail.google.com' },
+];
+
 /** encodeUriQuery encodes a string for use in a query parameter */
 function encodeUriQuery(val) {
   return encodeURIComponent(val).replace(/%20/g, '+');
 }
 
+/**
+ * Should we send a "badge" request to obtain the annotation count for the
+ * URL `uri`?
+ *
+ * @param {string} uri
+ * @return {boolean}
+ */
+function shouldQueryUri(uri) {
+  let url;
+
+  try {
+    url = new URL(uri);
+  } catch (e) {
+    return false;
+  }
+
+  // Make sure `uri` does not match ANY item in the blocklist
+  return BLOCKLIST.every(
+    blockedItem => url[blockedItem.urlProperty] !== blockedItem.value
+  );
+}
+
 /**
  * Queries the Hypothesis service that provides
  * statistics about the annotations for a given URL.
  */
 export function query(uri) {
   return fetch(settings.apiUrl + '/badge?uri=' + encodeUriQuery(uri), {
     credentials: 'include',
-  })
-    .then(function (res) {
-      return res.json();
-    })
-    .then(function (data) {
-      if (typeof data.total !== 'number') {
-        throw new Error('Annotation count is not a number');
-      }
-      return data;
-    });
+  }).then(res => res.json());
+}
+
+/**
+ * Retrieve the count of available annotations for `uri`
+ *
+ * @return {Promise<number>} - Annotation count for `uri`. Will be 0 if URI
+ *                             has a blocklist match or there are problems with
+ *                             the request to or response from the API
+ */
+export function getAnnotationCount(uri) {
+  const noValue = 0;
+  if (shouldQueryUri(uri)) {
+    return query(uri)
+      .then(data => {
+        if (!data || typeof data.total !== 'number') {
+          console.warn('Annotation count is not a number');
+          return noValue;
+        }
+        return data.total;
+      })
+      .catch(err => {
+        console.warn(`Failed to fetch annotation count for ${uri} (${err})`);
+        return noValue;
+      });
+  } else {
+    return Promise.resolve(noValue);
```

Comment:
This code was last touched before async/await was introduced. The minimum versions we support now though, according to `.babelrc`, are Chrome 64 and Firefox 68 which both support async/await natively. Therefore we could modernize this code to use async/await in future for readability.

Reply:
👍 I'll do this next time I alter this code, which I imagine might not be too long from now.